devito/finite_differences/tools.py

```diff
@@ -0,0 +1,211 @@
+            if any(len(expr_dict) > 1 for item in expr_dict):
+                raise NotImplementedError("Applying the chain rule to functions "
+                                          "with symbolic coefficients is not currently "
+                                          "supported")
+        kwargs['symbolic'] = expr._uses_symbolic_coefficients
+        return func(expr, *args, **kwargs)
+    return wrapper
+
+
+def dim_with_order(dims, orders):
+    """
+    Create all possible derivative order for each dims
+    for example dim_with_order((x, y), 1) outputs:
+    [(1, 0), (0, 1), (1, 1)]
+    """
+    ndim = len(dims)
+    max_order = np.max(orders)
+    # Get all combinations and remove (0, 0 ,0)
+    all_comb = tuple(product(range(max_order+1), repeat=ndim))[1:]
+    # Only keep the one with each dimension maximum order
+    all_comb = [c for c in all_comb if all(c[k] <= orders[k] for k in range(ndim))]
+    return all_comb
+
+
+def deriv_name(dims, orders):
+    name = []
+    for d, o in zip(dims, orders):
+        name_dim = 't' if d.is_Time else d.root.name
+        name.append('d%s%s' % (name_dim, o) if o > 1 else 'd%s' % name_dim)
+
+    return ''.join(name)
+
+
+def generate_fd_shortcuts(function):
+    """Create all legal finite-difference derivatives for the given Function."""
+    dimensions = function.indices
+    s_fd_order = function.space_order
+    t_fd_order = function.time_order if (function.is_TimeFunction or
+                                         function.is_SparseTimeFunction) else 0
+    orders = tuple(t_fd_order if i.is_Time else s_fd_order for i in dimensions)
+
+    from devito.finite_differences.derivative import Derivative
+
+    def deriv_function(expr, deriv_order, dims, fd_order, side=centered, **kwargs):
+        if isinstance(dims, (tuple, sympy.Tuple)):
+            return Derivative(expr, *dims, deriv_order=deriv_order, fd_order=fd_order,
+                              side=side, **kwargs)
+        return Derivative(expr, dims, deriv_order=deriv_order, fd_order=fd_order,
+                          side=side, **kwargs)
+
+    side = form_side(dimensions, function)
+    all_combs = dim_with_order(dimensions, orders)
+
+    derivatives = {}
+
+    # All conventiona fd shortcuts
+    for o in all_combs:
+        fd_dims = tuple(d for d, o_d in zip(dimensions, o) if o_d > 0)
+        d_orders = tuple(o_d for d, o_d in zip(dimensions, o) if o_d > 0)
+        fd_orders = tuple(t_fd_order if d.is_Time else s_fd_order for d in fd_dims)
+
+        deriv = partial(deriv_function, deriv_order=d_orders, dims=fd_dims,
+                        fd_order=fd_orders,
+                        stagger=tuple(side[d] for d in fd_dims))
+        name_fd = deriv_name(fd_dims, d_orders)
+        desciption = 'derivative of order %s w.r.t dimension %s' % (d_orders, fd_dims)
+        derivatives[name_fd] = (deriv, desciption)
+
+    # Add non-conventional, non-centered first-order FDs
```

Comment:
that's not complete, staggered is also handled inside

Reply:
Non-conventional ones yes, but all standard, including staggered, are handled in previous one.